### PR TITLE
Add API logic for create/delete message by adding a generic api endpoint for messages

### LINKED ISSUE #227: Add create and delete messages APIs
In this issue, we will try to add a generic endpoint in the backend that is able to accept the create/delete message request from the frontend .js file to finish the user interaction flow. 

## Justification
Per the way we structured our frontend and messages, we can use one endpoint, named ```MessageView```, as a generic entry point for create/delete requests.

- when a create comes, we will accept it, auto-assign it with an incremental token_id, take its null update_on field, and store it in database. Apparently as we have careful structured the message create logic, we need no additional changes/processing at this point
- when a delete request comes, we will simple accept it. The workflow will be the same as create, as basically we are not deleting anything but appending a new message that has a valid value in update_on in it with a null content that indicates this message is deleted

See issue #225 for further details about the message update model for the frontend and backend

## TODO:
- implement an entry point for the create/delete request from the frontend in the form of api
- update models.py to add auto-assign token_id feature, add update_on field
- implement more related codes for this api

### PR DESCRIPTION
**Pull request by MuyanXie**
PR for Add API logic for creating/delete messages by adding a generic API endpoint for messages

This PR is related to Issue #227 

TODO:

- Add a generic API entry point for creating and deleting messages
- due to the unique nature of our message fetch system, as long as we can implement an efficient lock system for updating our message token_id, we can keep the create/delete API in one single entry point
- build test case and stick to current format of the API team

Sticker:

Please use this API in later frontend page update with delete/create feature

### CODE REVIEW

**Review by jb-01 — APPROVED** [review]
This does a great job of creating generic API endpoint for creating and deleting chat messages! Incorporating auto-incrementing IDs is a great feature of this PR.

Please fix the merge conflicts with `dev` before closing it.

**Comment by MuyanXie** [comment]
Reminder: Will actually need to add tests for this API; Will re-request review after adding the tests

File: src/chigame/api/serializers.py [hunk 1]
```diff
@@ -30,3 +30,27 @@ class UserSerializer(serializers.ModelSerializer):
     class Meta:
         model = User
         fields = ("id", "name", "email")
+
+
+class MessageSerializer(serializers.ModelSerializer):
```
Comment (elizabethli31) [inline src/chigame/api/serializers.py:35]: I am getting this error when I try to post
image
I'm also seeing a 405 error
image

**Review by elizabethli31 — CHANGES_REQUESTED** [review]
Model change looks good, but I'm not sure what's going on with the api. Also, will you be adding the tests to this PR or another one?

File: src/chigame/api/serializers.py (hunk 1 above)
Comment (MuyanXie) [reply src/chigame/api/serializers.py:35]: For the DNE error, it seems like the POST request is sending a tournament ID that does not yet exist in the database. Please make sure that a valid tournament ID is provided and a chat is created for that tournament in the database before sending the request. We haven't yet included an error handling for this, as in our later PRs, we will auto-create a chat for every tournament once a tournament is created, and we will only send a request to this endpoint with a valid tournament ID from the front end. If you feel like we should have some error handling for this entry point, I'll definitely go ahead and add one.

For the 405 error, the error message you provided says "Method GET is not allowed"; for that endpoint, we only allow a POST method. Let's switch to the POST method and then see if the 405 error persists.

### Sample Post request body

__make sure you have a tournament with pk 1 and have created a chat for it__

{
    "sender": "admin@example.com",
    "content": "This is a test script 1",
    "update_on": null,
    "tournament": "1"
}

**Comment by MuyanXie** [comment]
> Model change looks good, but I'm not sure what's going on with the api. Also, will you be adding the tests to this PR or another one?

I will add tests to this PR to facilitate debugging and code review process

File: src/chigame/api/serializers.py (hunk 1 above)
Comment (MuyanXie) [reply src/chigame/api/serializers.py:35]: In later reviews, if we wish to test the API, we can also use the test_api.py file to test if the API is behaving logically:

```
python3 manage.py test chigame.api.tests.test_api.ChatTests
```

**Comment by MuyanXie** [comment]
Updated with Test API logics. We can now test API with 

```
python3 manage.py test chigame.api.tests.test_api.ChatTests
```

**Review by elizabethli31 — APPROVED** [review]
Tested again with a manually created chat, and it works! You've implemented very thorough tests which will be extremely helpful to build upon later on. LGTM!

**Comment by MuyanXie** [comment]
Closing Statement:

@MuyanXie worked on coding the API and coding the tests for the API. @MikeChen012345 worked on architecture the API endpoint as well as how the API should look like to work together with the frontend. @jb-01 worked on facilitating the review and debugging of a generalized entry point.

**Comment by elizabethli31** [comment]
Issue Score: Excellent

Comments:
Nice addition of tests along with the other mentioned tasks as part of this PR.

**Pull request merged**